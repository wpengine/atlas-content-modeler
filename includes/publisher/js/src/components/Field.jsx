import React, { useState, useRef } from "react";
import MediaUploader from "./MediaUploader";
import RichText from "./RichText";
import Relationship from "./relationship";
import Text from "./Text";
import Email from "./Email";
import Number from "./Number";
import Date from "./Date";
import Icon from "acm-icons";
import { sprintf, __ } from "@wordpress/i18n";

const defaultError = "This field is required";
const integerRegex = /^[-+]?\d+/g;
const decimalRegex = /^\-?(\d+\.?\d*|\d*\.?\d+)$/g;

export default function Field(props) {
	const { field, modelSlug, first } = props;
	const [errors, setErrors] = useState({});

	/**
	 * Adjusts the custom error feedback messages displayed below fields based
	 * on the HTML5 validation failure type.
	 *
	 * @param {object} event onChange
	 * @param {object} field
	 */
	function validate(event, field) {
		let error = defaultError;

		if (field.type === "number") {
			if (!integerRegex.test(event.target.value)) {
				error = __(
					"Value must be an integer.",
					"atlas-content-modeler"
				);
			}

			if (field.numberType === "decimal") {
				if (!decimalRegex.test(event.target.value)) {
					error = __(
						"Value must be a decimal.",
						"atlas-content-modeler"
					);
				}
			}

			if (event.target.validity.rangeOverflow) {
				error = sprintf(
					__("Maximum value is %s.", "atlas-content-modeler"),
					event.target.max.toString()
				);
			} else if (event.target.validity.rangeUnderflow) {
				error = sprintf(
					__("Minimum value is %s.", "atlas-content-modeler"),
					event.target.min.toString()
				);
			} else if (event.target.validity.stepMismatch) {
				error = sprintf(
					__(
						"Value must be a multiple of %s.",
						"atlas-content-modeler"
					),
					event.target.step.toString()
				);
			}
		}

		if (field.type === "text") {
			if (event.target.validity.tooShort) {
				error = sprintf(
					__("Minimum length is %d.", "atlas-content-modeler"),
					event.target.minLength
				);
			} else if (event.target.validity.tooLong) {
				error = sprintf(
					__("Maximum length is %d.", "atlas-content-modeler"),
					event.target.maxLength
				);
			}
		}

		if (field.type === "email") {
			if (!event.target.validity.valid) {
				error = __("Value must be an email.", "atlas-content-modeler");
			}

			if (field?.required && event.target.value.trim() === "") {
				error = defaultError;
			}

			if (
				field?.allowedDomains &&
				event.target.validity.patternMismatch
			) {
				error = __(
					"Email must end with an allowed domain.",
					"atlas-content-modeler"
				);
			}
		}

		setErrors({ ...errors, [field.slug]: error });
	}

	return (
		<>
			<div
				id={`field-${field.slug}`}
				className={`field d-flex flex-column ${field.type}`}
				{...(first && { "data-first-field": true })}
			>
				{fieldMarkup(field, modelSlug, errors, validate)}
			</div>
		</>
	);
}

function fieldMarkup(field, modelSlug, errors, validate) {
	switch (field.type) {
		case "relationship":
			return (
				<Relationship
					field={field}
					modelSlug={modelSlug}
					required={field.required}
				/>
			);
		case "media":
			return (
				<MediaUploader
					field={field}
					modelSlug={modelSlug}
					required={field.required}
				/>
			);
		case "text":
			return (
				<Text
					field={field}
					modelSlug={modelSlug}
					errors={errors}
					validate={validate}
					defaultError={defaultError}
				/>
			);
		case "number":
			return (
				<Number
					field={field}
					modelSlug={modelSlug}
					errors={errors}
					validate={validate}
					defaultError={defaultError}
				/>
			);
		case "date": // @todo split this out for proper browser and datepicker support
			return (
				<Date
					field={field}
					modelSlug={modelSlug}
					defaultError={defaultError}
				/>
			);

		case "richtext":
			return (
				<RichText
					field={field}
					modelSlug={modelSlug}
					errors={errors}
					validate={validate}
					defaultError={defaultError}
				/>
			);

		case "boolean":
			const [checked, setChecked] = useState(field.value === "on");
			return (
				<>
					{field?.description && (
						<p className="help mb-3">{field.description}</p>
					)}
					<label
						className="check-container"
						htmlFor={`atlas-content-modeler[${modelSlug}][${field.slug}]`}
					>
						{field.name}
						<input
							type="checkbox"
							name={`atlas-content-modeler[${modelSlug}][${field.slug}]`}
							id={`atlas-content-modeler[${modelSlug}][${field.slug}]`}
							checked={checked}
							onChange={(event) => setChecked(!checked)}
							required={field.required}
						/>
						<span className="error">
							<Icon type="error" />
							<span role="alert">{defaultError}</span>
						</span>
						{/* span is used for custom checkbox styling purposes */}
						<span className="checkmark"></span>
					</label>
				</>
			);

		case "multipleChoice":
			if (field.listType === "multiple") {
				return (
					<fieldset>
						<legend>{field.name}</legend>
						{field?.description && (
							<p className="help mb-0">{field.description}</p>
						)}
						{field.choices.map((item, index) => {
							return (
								<label
									key={index}
									className="check-container multi-check-container"
									htmlFor={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}][${item.slug}]`}
								>
									{item.name}
									<input
										type="checkbox"
										name={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}][${item.slug}]`}
										id={`atlas-content-modeler[${modelSlug}][${field.slug}][${index}][${item.slug}]`}
										placeholder="Option Name"
										value={item.slug}
										defaultChecked={
											field.value &&
											field.value.some(
												(slug) => slug === item.slug
											)
										}
									/>
									<span className="error">
										<Icon type="error" />
										<span role="alert">{defaultError}</span>
									</span>
									{/* span is used for custom checkbox styling purposes */}
									<span className="checkmark"></span>
								</label>
							);
						})}
					</fieldset>
				);
			}
			if (field.listType === "single") {
				return (
					<fieldset>
						<legend>{field.name}</legend>
						{field?.description && (
							<p className="help mb-0">{field.description}</p>
						)}
						{field.choices.map((item, index) => {
							return (
								<label className="radio-container" key={index}>
									{item.name}
									<input
										type="radio"
										name={`atlas-content-modeler[${modelSlug}][${field.slug}]`}
										id={`atlas-content-modeler[${modelSlug}][${field.slug}][${item.name}]`}
										value={item.slug}
										defaultChecked={
											field.value[0] === item.slug
										}
									/>
									<span className="error">
										<Icon type="error" />
										<span role="alert">{defaultError}</span>
									</span>
									{/* span is used for custom radio styling purposes */}
									<span className="radio-select"></span>
								</label>
							);
						})}
					</fieldset>
				);
			}
		case "email":
			return (
				<Email
					field={field}
					modelSlug={modelSlug}
					errors={errors}
					validate={validate}
					defaultError={defaultError}
				/>
			);
		default:
			return `TODO: ${field.type} fields`;
	}
}
