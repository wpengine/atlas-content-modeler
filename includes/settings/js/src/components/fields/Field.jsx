import React, { useState, useContext } from "react";
import { Draggable } from "react-beautiful-dnd";
import Form from "./Form";
import Icon from "../../../../../components/icons";
import supportedFields from "./supportedFields";
import { FieldOptionsDropdown } from "./FieldOptionsDropdown";
import { ModelsContext } from "../../ModelsContext";
import FieldButtons from "../FieldButtons";
import { sprintf, __ } from "@wordpress/i18n";

const getItemStyle = (isDragging, draggableStyle) => {
	return {
		...draggableStyle,
	};
};

function Field({
	data = {},
	editing,
	id,
	index,
	model,
	open = false,
	position,
	positionAfter,
	setInfoTag,
	promptToSaveChanges,
	hasDirtyField,
	type = "text",
}) {
	const [activeForm, setActiveForm] = useState(type);
	const { dispatch } = useContext(ModelsContext);

	// Closed fields appear as a row with a summary of info.
	if (!open) {
		const typeLabel = supportedFields[data.type];
		const reorderInfoTag = (
			<>
				Press space to begin reordering the “{data.name}” field. Use{" "}
				<Icon type="downarrow" /> and <Icon type="uparrow" /> keys to
				reorder, then space to finish and save.
			</>
		);
		return (
			<Draggable
				key={id}
				draggableId={String(id)}
				index={index}
				disableInteractiveElementBlocking={true}
			>
				{(provided, snapshot) => (
					<div
						className="draggable"
						ref={provided.innerRef}
						{...provided.draggableProps}
						style={getItemStyle(
							snapshot.isDragging,
							provided.draggableProps.style
						)}
					>
						<li
							key={id}
							className={`${
								snapshot.isDragging
									? "closed dragging"
									: "closed"
							} d-flex flex-column d-sm-flex flex-sm-row justify-content-start`}
						>
							<div className="reorder order-2 order-sm-0">
								<button
									{...provided.dragHandleProps}
									onFocus={() => {
										setInfoTag(reorderInfoTag);
									}}
									onBlur={() => setInfoTag(null)}
								>
									<Icon type="reorder" />
								</button>
							</div>
							<button
								className="edit order-sm-2 order-1 d-flex flex-column d-sm-flex flex-sm-row ml-sm-4 flex-grow-1"
								onClick={() => {
									if (!promptToSaveChanges()) {
										dispatch({
											type: "openField",
											id: data.id,
											model: model.slug,
										});
									}
								}}
								aria-label={`Edit the ${data.name} field`}
							>
								<div
									className="row w-100"
									style={{ "--bs-gutter-x": "0em" }}
								>
									<div className="col-sm-3 col-12 text-center text-sm-start ms-3 ms-sm-0">
										<div className="type mb-3 mb-sm-0">
											<Icon type={data.type} />
											{typeLabel}
										</div>
									</div>
									<div className="col-sm-3 col-12 text-center text-sm-start">
										<div className="widest mb-2 mb-sm-0">
											<strong>{data.name}</strong>
											{"  "}
											<span
												title={__(
													"API Identifier",
													"atlas-content-modeler"
												)}
											>
												({data.slug})
											</span>
										</div>
									</div>
									<div className="col-sm-6 col-12 text-center text-sm-end">
										<div className="tags">
											{data?.isTitle && (
												<span className="tag tag-title">
													entry title
												</span>
											)}
										</div>
									</div>
								</div>
							</button>
							<div className="order-0 my-2 my-sm-0 text-end order-sm-2 ms-sm-auto">
								<FieldOptionsDropdown
									field={data}
									model={model}
								/>
							</div>
						</li>
						<li
							className={
								snapshot.isDragging
									? "add-item dragging"
									: "add-item"
							}
						>
							<button
								onClick={() => {
									if (!promptToSaveChanges()) {
										dispatch({
											type: "addField",
											position: positionAfter,
											model: model.slug,
										});
									}
								}}
								aria-label={`Add a new field below the ${data.name} ${data.type} field`}
							>
								<Icon type="add" size="large" />
							</button>
						</li>
					</div>
				)}
			</Draggable>
		);
	}

	const formFieldTitle = editing
		? `“${data.name}”`
		: supportedFields[activeForm];

	// Open fields appear as a form to enter or edit data.
	return (
		<li className="open-field" key={id}>
			{!editing && (
				<FieldButtons
					activeButton={activeForm}
					clickAction={setActiveForm}
				/>
			)}
			<div className={editing ? "field-form editing" : "field-form"}>
				<div className="d-flex flex-row">
					<div>
						<h3>
							{editing
								? __("Editing", "atlas-content-modeler")
								: __("New", "atlas-content-modeler")}{" "}
							{formFieldTitle} Field
						</h3>
					</div>

					{editing && (
						<div className="ms-auto">
							<FieldOptionsDropdown field={data} model={model} />
						</div>
					)}
				</div>
				<Form
					type={activeForm}
					storedData={data}
					editing={editing}
					id={id}
					position={position}
					hasDirtyField={hasDirtyField}
				/>
			</div>
		</li>
	);
}

export default Field;
