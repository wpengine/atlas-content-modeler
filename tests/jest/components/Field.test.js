import React from "react";
import { act, create } from "react-test-renderer";
import Field from "../../../includes/content-editing/js/src/components/Field";

const model = {
	slug: "geese",
};

const textField = {
	name: "First Name",
	slug: "firstName",
	type: "text",
	value: "John",
};

const numberField = {
	name: "Age",
	slug: "age",
	type: "number",
	value: "100",
};

describe("Field", () => {
	let root;

	it("renders a proper text field", () => {
		act(() => {
			root = create(<Field field={textField} modelSlug={model.slug} />);
		});

		expect(root.toJSON()).toMatchSnapshot();
	});

	it("renders a proper number field", () => {
		act(() => {
			root = create(<Field field={numberField} modelSlug={model.slug} />);
		});

		expect(root.toJSON()).toMatchSnapshot();
	});
});
