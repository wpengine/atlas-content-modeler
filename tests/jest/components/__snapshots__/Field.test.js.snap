// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field renders a proper number field 1`] = `
<div
  className="field number"
>
  <label
    htmlFor="wpe-content-model[geese][age]"
  >
    Age
  </label>
  <br />
  <input
    defaultValue="100"
    id="wpe-content-model[geese][age]"
    name="wpe-content-model[geese][age]"
    type="number"
  />
</div>
`;

exports[`Field renders a proper text field 1`] = `
<div
  className="field text"
>
  <label
    htmlFor="wpe-content-model[geese][firstName]"
  >
    First Name
  </label>
  <br />
  <input
    defaultValue="John"
    id="wpe-content-model[geese][firstName]"
    name="wpe-content-model[geese][firstName]"
    type="text"
  />
</div>
`;
